Log in log out

wilsonyqm opened the pull request:
Closes #11

wilsonyqm commented:
Hi Daniel, please take a look if you have time. Thanks @d-ray

d-ray commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  end
+
+  def create
+    @user = User.new(user_params)    # Not the final implementation!
```
Is this comment still accurate?

wilsonyqm replied:
Deleted

d-ray commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,30 @@
+class UsersController < ApplicationController
+  before_action :no_login, only: [:create, :new]
+  include SessionsHelper
```
Including another controller's helper is an antipattern.  How could you restructure to provide the methods you need here without including the entire `SessionsHelper`?

wilsonyqm replied:
The method that has been used on both controller can be moved to the applcation_helper.rb

rmacklin replied:
The main thing here is that Helpers in rails have a special meaning. They are automatically mixed into the view context so methods can be called directly from the views. Since that's the case, if you need to use some methods in both a controller and a view, it's better to put those methods in another module (which is not a Helper) and include _that_ module in the Controller and in the Helper.

d-ray commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,39 @@
+    user = User.find_by(email: params[:session][:email])
+    if user.present? && user.authenticate(params[:session][:password])
+      log_in user
+      params[:session][:remember_me] == '1' ? remember(user) : forget
```
Do you need to actively `forget` if the user chooses not to be remembered?

wilsonyqm replied:
I think so, since forget here is to clear the cookies, if user chooses not to remember it, then there should be no cookie in the browser to remember it.

d-ray replied:
Is it possible to hit this action if the remember cookie is set?

wilsonyqm replied:
No, if remember cookie is set, you have to be in the logged in status. I have `before_filter :set_current_user` in applicationController.rb, which will check whether there is a cookie set, if yes then log in this user. 
`before_action :no_login, only: [:create, :new]` this line guarantees that you can not create a now session when you already logged in.

d-ray replied:
Right - not sure it was clear, but that was a followup to my original question.  The idea being that if (as you noted) you can't hit the this action when the remember cookie is set, there's no reason to clear the cookie when the user chooses not to be remembered.

wilsonyqm replied:
Yep, you are right. The forget here is kind of redundant, I will delete it.

d-ray commented on app/helpers/sessions_helper.rb:
```diff
@@ -0,0 +1,36 @@
+module SessionsHelper
+
+  def log_in(user)
```
Putting methods in a helper exposes them in the view layer; does it make sense to have all of these method available in there?

wilsonyqm replied:
made the forget and current_user method private.

d-ray commented on app/helpers/users_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module UsersHelper
```
Is there value in defining an empty helper?

wilsonyqm replied:
Deleted

wilsonyqm commented:
Some redundant code been deleted, and some comments are left to be discussed.

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -95,4 +99,47 @@ class ImagesCrudTest < FlowTestCase
     assert_equal 3, images_index_page.images.count
   end
+
+  test 'sign up' do
```
These flow tests are not part of "image CRUD" so they should be moved to a new file, e.g. `login_and_registration_test.rb`

wilsonyqm replied:
Moved

rmacklin commented on test/page_objects/images/index_page.rb:
```diff
@@ -10,6 +10,16 @@ def view!
         end
       end
 
+      def sign_up_log_in!
```
`log_in_or_sign_up!` would be a more appropriate name for this.

rmacklin replied:
Also, do we want to only allow users to log in or sign up from the images index page?

wilsonyqm replied:
Good point, I would move it to the Nav bar.

rmacklin commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,39 @@
+
+  private
+
+  def no_login
```
A better name for this would be `redirect_if_already_logged_in`

wilsonyqm replied:
Fixed

rmacklin commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,39 @@
+      log_out
+      flash[:success] = 'You successfully logged out'
+    else
+      flash[:danger] = 'You are already logged out'
```
Minor, but consider using `flash[:info]` here, since this isn't really "dangerous"

wilsonyqm replied:
Changed

rmacklin commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,39 @@
+
+  def no_login
+    if logged_in?
+      flash[:danger] = 'Please log out before you log in'
```
Maybe it'd be better to use:
```ruby
flash[:info] = 'You are already logged in'
```
which parallels line 26

wilsonyqm replied:
Modified

rmacklin commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,39 @@
+class SessionsController < ApplicationController
+  include SessionsHelper
```
Which methods does this controller need from this module? It would be better to put them in a new module (which is not a Helper) and include that module here and in SessionsHelper.

rmacklin commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,31 @@
+
+  private
+
+  def no_login
```
Again, `redirect_if_already_logged_in` would be a better name here.

wilsonyqm replied:
FIxed https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36/commits/10b1b87e2e787e2757f9d7eba7d76483e0a5c90d

rmacklin commented on app/models/user.rb:
```diff
@@ -0,0 +1,10 @@
+  VALID_EMAIL_FORMAT = /\A([\w+\-]\.?)+@[a-z\d\-]+(\.[a-z]+)*\.[a-z]+\z/i
+
+  validates :name,  presence: true, length: { maximum: 50 }
+  validates :email, presence: true, length: { maximum: 255 }, uniqueness: { case_sensitive: false },
```
I don't think you need `case_sensitive: false` given that your forcing the emails to be lowercase on line 2.

d-ray replied:
It's probably needed currently, since validations will run before the `before_save` hook (http://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html).  You could use a `before_validation` hook instead and then remove the `case_sensitive` option.

wilsonyqm replied:
Cool, modified with `before_validation`

rmacklin commented on app/models/user.rb:
```diff
@@ -0,0 +1,10 @@
+  validates :name,  presence: true, length: { maximum: 50 }
+  validates :email, presence: true, length: { maximum: 255 }, uniqueness: { case_sensitive: false },
+    format: { with: VALID_EMAIL_FORMAT }
+  has_secure_password
```
Can you move this below all the validations?

wilsonyqm replied:
Moved

rmacklin commented on app/models/user.rb:
```diff
@@ -0,0 +1,10 @@
+class User < ActiveRecord::Base
+  before_save { self.email = email.downcase }
+  VALID_EMAIL_FORMAT = /\A([\w+\-]\.?)+@[a-z\d\-]+(\.[a-z]+)*\.[a-z]+\z/i
```
Is this the same regex used for the share image form? If so, do you think it's a good idea to duplicate it? Moreover, can you think of a better way to DRY email validation? (If you get stuck, talk to Xin)

wilsonyqm replied:
Fixed with `ImagesController::ShareForm::VALID_EMAIL`

Xin-Tong replied:
@wilsonyqm There is a gem https://github.com/validates-email-format-of/validates_email_format_of. You can use this.

rmacklin commented on app/views/images/index.html.erb:
```diff
@@ -1,13 +1,24 @@
+  <div id="images_list" class="row">
+    <% @images.each do |image| %>
+      <div class="js-image-card col-sm-6 col-xs-12">
+        <div class="card-block img-fluid", data-image-id=<%= image.id %>>
```
How is this invalid HTML still here❓❗ I filed https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/issues/41. Please do that ASAP.

wilsonyqm replied:
Fixed

rmacklin commented on app/views/images/index.html.erb:
```diff
@@ -1,13 +1,24 @@
+        <%= link_to 'Log Out', logout_path, method: :delete, class: 'btn btn-warning pull-xs-left' %>
+      <% else %>
+        <%= link_to 'Log In / Sign Up', login_path, class: 'btn btn-info pull-xs-left' %>
+      <% end %>
```
Do we want to only allow users to log in/sign up/log out from the images index page?

wilsonyqm replied:
Moved to Navbar

rmacklin commented on app/views/images/index.html.erb:
```diff
@@ -1,13 +1,24 @@
-<%= link_to 'Insert Image', new_image_path, class: 'btn btn-success pull-xs-right' %>
 
 <h1 class="heading">Listing of Images</h1>
+<div>
```
You should never have a `div` with no attributes.

rmacklin commented on app/views/sessions/new.html.erb:
```diff
@@ -0,0 +1,11 @@
+  <div class="col-md-6 col-md-offset-3">
+    <%= simple_form_for(:session, url: login_path, html: { class: 'js-session-form' } ) do |f| %>
+      <%= f.input :email%>
+      <%= f.input :password%>
```
Please put a space before the closing `%>`s

wilsonyqm replied:
Added

rmacklin commented on app/views/sessions/new.html.erb:
```diff
@@ -0,0 +1,11 @@
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= simple_form_for(:session, url: login_path, html: { class: 'js-session-form' } ) do |f| %>
```
Please either remove the parentheses or remove the space before the closing parenthesis.

rmacklin replied:
Also, I don't think you need to give this a `js-` class. You can select it by its `id`.

wilsonyqm replied:
`as: :session_form` does not work here, it won't generate an id. I prefer this way

rmacklin replied:
You can manually specify an `id` in the `html` hash instead of specifying a `class`

rmacklin commented on app/views/users/new.html.erb:
```diff
@@ -0,0 +1,13 @@
+<% provide(:title, 'Sign up') %>
```
What's the purpose of this line?

wilsonyqm replied:
Deleted, it's from the tutorial.

rmacklin commented on app/views/users/new.html.erb:
```diff
@@ -0,0 +1,13 @@
+<h1>Sign up here</h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= simple_form_for(@user, html: { class: 'js-user-form'} ) do |f| %>
```
Same as above: Please either remove the parentheses or remove the space before the closing parenthesis.

rmacklin replied:
Also, I don't think you need to give this a `js-` class. You can select it by its `id`.

wilsonyqm replied:
we can use `as: :user_form` here, since user is a class been define, but session is not a defined class.

rmacklin commented on db/migrate/20160426211440_add_index_to_users_email.rb:
```diff
@@ -0,0 +1,5 @@
+class AddIndexToUsersEmail < ActiveRecord::Migration
```
You can combine this with the previous migration.

rmacklin commented on db/migrate/20160426212826_add_password_digest_to_users.rb:
```diff
@@ -0,0 +1,5 @@
+class AddPasswordDigestToUsers < ActiveRecord::Migration
```
You can combine this with the previous migration.

wilsonyqm replied:
I deleted this file, which is useless in this case, we do not have password_digest anymore.

rmacklin commented on db/schema.rb:
```diff
@@ -21,23 +21,31 @@
-    t.string   "taggable_type",             null: false
+    t.integer  "tag_id"
+    t.integer  "taggable_id"
+    t.string   "taggable_type"
```
Why did all these `null: false`s disappear?

wilsonyqm replied:
I run db:migrate, and they disapeared

rmacklin replied:
That's not good. That makes me thing your migrations are not correct.

rmacklin replied:
More generally, you should pay very close attention to changes to `schema.rb`, as these are super important.

rmacklin commented on db/migrate/20160426204118_create_users.rb:
```diff
@@ -0,0 +1,10 @@
+  def change
+    create_table :users do |t|
+      t.string :name
+      t.string :email
```
Can these be null?

wilsonyqm replied:
No

rmacklin replied:
Right, so we should add `null: false`

rmacklin commented on test/controllers/sessions_controller_test.rb:
```diff
@@ -0,0 +1,91 @@
+require 'test_helper'
+
+class SessionsControllerTest < ActionController::TestCase
+  include SessionsHelper
```
As mentioned above, we shouldn't include a view helper here.

wilsonyqm replied:
Fixed

rmacklin commented on test/controllers/sessions_controller_test.rb:
```diff
@@ -0,0 +1,91 @@
+
+  test 'new' do
+    get :new
+    assert_response :success
```
What else can you assert here?

wilsonyqm replied:
Added assert session new form

rmacklin commented on test/controllers/sessions_controller_test.rb:
```diff
@@ -0,0 +1,91 @@
+    assert_equal 'Invalid password/username', flash[:danger]
+  end
+
+  test 'create with missing user' do
```
This is "missing email" rather than "missing user" (which sounds like "nonexistent user", which is tested above).

wilsonyqm replied:
Ok

rmacklin commented on test/controllers/sessions_controller_test.rb:
```diff
@@ -0,0 +1,91 @@
+    post :create, session: { email: username, password: password }
+    assert_equal user.id, session[:user_id]
+    assert_equal flash[:success], 'Welcome to the Share Image App!'
+    assert_redirected_to images_path
```
Can you move this, and all other response status assertions to directly after the line that makes the request?

wilsonyqm replied:
moved

rmacklin commented on test/controllers/sessions_controller_test.rb:
```diff
@@ -0,0 +1,91 @@
+    username = 'username@email.com'
+    password = 'passwd'
+
+    post :create, session: { email: username, password: password}
```
👾

wilsonyqm replied:
Fixed.

rmacklin commented on test/page_objects/images/signup_page.rb:
```diff
@@ -0,0 +1,25 @@
+  module Users
+    class SignupPage < PageObjects::Document
+      path :signup
+      path :new_user
```
Do you think it's necessary for our app to provide both `/signup` and `/users/new`? Maybe we only need to provide one.

wilsonyqm replied:
how to do that at routes.rb?

rmacklin replied:
What have you tried?

wilsonyqm replied:
I changed `  resources :users, only: [:create, :new]` to `  resources :users, only: [:create]`

rmacklin replied:
That should work. However, using only a single action in the `only` option of a `resources` declaration is indicative that it's not really a full-blown resource (so using `resources` may not be appropriate). Furthermore, declaring it as `resources :users, only: [:create]` would mean the path for `users#create` is a `POST` to `/users`. With that in mind, you could delete the `resources :users, only: [:create]` line altogether, and replace it with:
```ruby
post   'signup'  => 'users#create'
```
which would create a nice symmetry with the login routes:
```ruby
get    'signup'  => 'users#new'
post   'signup'  => 'users#create'
get    'login'   => 'sessions#new'
post   'login'   => 'sessions#create'
```

wilsonyqm replied:
Make sense.

rmacklin commented on test/page_objects/images/signup_page.rb:
```diff
@@ -0,0 +1,25 @@
+        element :password_confirmation
+      end
+
+      def create_user!(name: nil, email: nil, password: nil, password_confirmation: nil)
```
`create_account!` or `create_my_account!` would be a better name for this, as it matches the terminology users of our application would think in ("I'm creating an account") rather than the terminology that corresponds to how our application is implemented ("we're creating a `User` model"). This point is driven further by the fact that the button text is "Create my account".

wilsonyqm replied:
Got you

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -95,4 +99,47 @@ class ImagesCrudTest < FlowTestCase
+    images_index_page = PageObjects::Images::IndexPage.visit
+
+    user_login_page = images_index_page.sign_up_log_in!
+    user_signup_page = user_login_page.sign_up!
```
You can actually combine these:
```ruby
user_signup_page = images_index_page.sign_up_log_in!.sign_up!
```

wilsonyqm replied:
Fixed

wilsonyqm commented:
Squashed the commits. Make minor format changes according to pronto.

rmacklin commented on lib/session_methods.rb:
```diff
@@ -0,0 +1,37 @@
+
+  def remember(user)
+    cookies.permanent.signed[:user_id] = user.id
+  end
```
This method is not needed by everywhere that includes `SessionMethods`

wilsonyqm replied:
Moved to session_controller as private.

wilsonyqm replied:
Moved

rmacklin commented on lib/session_methods.rb:
```diff
@@ -0,0 +1,37 @@
+    !current_user.nil?
+  end
+
+  def log_out
```
This method is does not need to be shared via the `SessionMethods` mixin. Please move it to where it is used.

wilsonyqm replied:
Moved

wilsonyqm replied:
Moved

rmacklin commented on lib/session_methods.rb:
```diff
@@ -0,0 +1,37 @@
+
+  private
+
+  def forget
```
Same as above

wilsonyqm replied:
Moved

wilsonyqm replied:
Moved

rmacklin commented on lib/session_methods.rb:
```diff
@@ -0,0 +1,37 @@
+module SessionMethods
```
`____Methods` is a poor name for modules. We should avoid using it. Consider `SessionUtilities`, though there might be an even better name.

wilsonyqm replied:
Changed

rmacklin commented on app/views/sessions/new.html.erb:
```diff
@@ -0,0 +1,13 @@
+  <div class="col-md-6 col-md-offset-3">
+    <%= simple_form_for :session,
+                        url: login_path,
+                        html: { class: 'js-session-form', novalidate: '' } do |f| %>
```
Please use an `id` instead of a `js-` class. https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36#discussion_r62400724

wilsonyqm replied:
replaced.

rmacklin commented on app/views/users/new.html.erb:
```diff
@@ -0,0 +1,15 @@
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= simple_form_for @user,
+                        as: :user,
```
If you remove this line, does it still work?

wilsonyqm replied:
replaced with id in html block

rmacklin replied:
That's not what I meant. I think it should give a default `id` without you specifying one.

rmacklin commented on db/migrate/20160426211440_add_index_to_users_email.rb:
```diff
@@ -0,0 +1,6 @@
+class AddIndexAndColumnToUsers < ActiveRecord::Migration
```
You can still combine this with the previous migration https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36#discussion_r62155207

wilsonyqm replied:
combined

rmacklin commented on test/controllers/sessions_controller_test.rb:
```diff
@@ -0,0 +1,109 @@
+  test 'new' do
+    get :new
+    assert_select '.js-session-form', 1
+    assert_response :success
```
Generally, the response status assertions should be the first assertion after the request is made.

wilsonyqm replied:
Moved

rmacklin commented on app/mailers/image_mailer.rb:
```diff
@@ -1,9 +1,10 @@
+  def send_email(image, share_form, user=nil)
     @recipient = share_form.recipient
+    @sender = user.present? ? user.email : 'wilsonyqm@gmail.com'
     @subject = share_form.subject.presence || 'Your friend sent you this image'
```
This is pre-existing but `@subject` and `@recipient` do not need to be instance variables.

wilsonyqm replied:
changed to local variable

rmacklin commented on app/mailers/image_mailer.rb:
```diff
@@ -1,9 +1,10 @@
     @url = image.url
     @content = share_form.content
-    mail(to: @recipient, subject: @subject)
+    mail(to: @recipient, subject: @subject, from: @sender)
```
It's a bit sketchy to impersonate the sender here. I think instead of this, you could simply modify the mailer to mention the name and email address of the logged in user.

wilsonyqm replied:
removed the @sender, changed the mail text

rmacklin commented on db/migrate/20160426204118_create_users.rb:
```diff
@@ -0,0 +1,10 @@
+class CreateUsers < ActiveRecord::Migration
+  def change
+    create_table :users do |t|
+      t.string :name
```
Can this be null?

wilsonyqm replied:
Added

rmacklin commented on app/controllers/sessions_controller.rb:
```diff
@@ -28,6 +28,15 @@ def destroy
   end
 
   private
+  def log_out
```
Please add a newline above this line

wilsonyqm replied:
Added

wilsonyqm commented:
I have addressed the comments above in commit https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36/commits/5ec1983dd22375d4df2e9ac307a0b68b90b3173d

rmacklin commented on app/mailers/image_mailer.rb:
```diff
@@ -2,9 +2,9 @@ class ImageMailer < ApplicationMailer
   def send_email(image, share_form, user=nil)
     @recipient = share_form.recipient
     @sender = user.present? ? user.email : 'wilsonyqm@gmail.com'
```
I don't think this should have a default which is your email. If `user` is `nil`, the email should just say "Somebody sent you an image". But if `user` is provided, it should say "Foo Bar (foo@bar.com) sent you an image". So, to support this, you could write: 
```ruby
@sender = user.present? "#{user.name} (@{user.email)}" : 'Somebody'
```
and then, in the mailer template, you'd write:
```erb
<%= @sender %> sent you an image.
```
In all cases, the email will still be "from" our application (similar to how you'd receive an email from Facebook or Twitter on behalf of someone else)

wilsonyqm replied:
Fixed.

rmacklin commented on app/views/image_mailer/send_email.html.erb:
```diff
@@ -1,5 +1,5 @@
 <p>
-  Hi <%= @recipient %>, This is a message with an image<br>
+  Hi <%= @recipient %>, This is a message <%= "from #{@sender} " if @sender.present? %>with an image<br>
```
The way your code is written, `@sender` is always `present?`. But, I suggest changing it to https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36#discussion_r63062853

wilsonyqm replied:
Fixed.

rmacklin commented on app/views/sessions/new.html.erb:
```diff
@@ -2,7 +2,7 @@
     <%= simple_form_for :session,
                         url: login_path,
-                        html: { class: 'js-session-form', novalidate: '' } do |f| %>
+                        html: { id: 'session-form', novalidate: '' } do |f| %>
```
Our convention is that `id`s, unlike `class`es, should use `_` (not `-`) for the word delimiter. So this should be `session_form`.

wilsonyqm replied:
changed.

rmacklin commented on app/views/users/new.html.erb:
```diff
@@ -2,9 +2,8 @@
-                        as: :user,
                         url: signup_path,
-                        html: { novalidate: '' } do |f| %>
+                        html: { id: 'user-form', novalidate: '' } do |f| %>
```
If you don't specify an `id` will there be a default one? See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36#discussion_r63061294

wilsonyqm replied:
it's new_user as default id

rmacklin commented on db/migrate/20160426204118_create_and_modify_users.rb:
```diff
@@ -0,0 +1,11 @@
+class CreateAndModifyUsers < ActiveRecord::Migration
```
This should just be `CreateUsers`. The users table did not exist prior to this migration, so there's nothing being "modified".

rmacklin commented:
See new comments.

wilsonyqm commented:
Addressed in https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36/commits/ba7e32788e43c9e23de149b7d6c3414d5a847059

rmacklin commented on app/views/layouts/_navbar.html.erb:
```diff
@@ -1,13 +1,17 @@
+      <% if @current_user.present? %>
+        <%= link_to 'Log Out', logout_path, method: :delete, class: 'btn btn-warning pull-xs-left' %>
+      <% else %>
+        <%= link_to 'Log In / Sign Up', login_path, class: 'btn btn-info pull-xs-left' %>
```
The other nav bar items don't use bootstrap button styles. You may want to match the style of the rest of the nav bar items.

wilsonyqm replied:
Fixed by using nav_item function, fix the style

rmacklin commented on config/application.rb:
```diff
@@ -41,5 +41,7 @@ class Application < Rails::Application
 
     ActionMailer::Base.delivery_method = :smtp
+
+    config.autoload_paths += ["#{config.root}/lib"]
```
Instead of doing this and putting `SessionUtitlities` in `lib`, you could put it in `controllers/concerns`

wilsonyqm replied:
moved.

rmacklin commented:
This has not been addressed: https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36#discussion_r63063662

rmacklin commented on test/flow_tests/login_and_registration_test.rb:
```diff
@@ -0,0 +1,62 @@
+    assert_equal 'Welcome to the Share Image App!', images_index_page.flash_message(:success)
+  end
+
+  test 'logout' do
```
Since this test needs to first login, you could just combine this and the previous test into a `'log in and log out'` test case

wilsonyqm replied:
Combined

rmacklin commented on test/mailers/image_mailer_test.rb:
```diff
@@ -73,7 +75,43 @@ class ImageMailerTest < ActionMailer::TestCase
     end
   end
+
+  test 'share image when logged in' do
```
This is really `'share image from a User'`. As far as the mailer is concerned, it just gets passed a `User` instance. Only the controller knows that that happens to be the logged in user.

But...this makes me realize there should be another controller test case to make sure it passes the logged in user to the mailer.

wilsonyqm replied:
This is more likely an integration test rather than a controller test. Do you think it's necessary to write this test in controller test?

rmacklin replied:
No, it's not. It's still testing a single action. And yes, I think you should verify that the correct user is passed to the mailer.

rmacklin commented on test/page_objects/images/login_page.rb:
```diff
@@ -0,0 +1,24 @@
+module PageObjects
+  module Sessions
+    class LoginPage < PageObjects::Document
```
I think the `Sessions` namespace is unnecessary. From the users perspective, this page is at `/login` and is just the "login page" not the "sessions login page". The fact that we hit the sessions controller is an implementation detail hidden from our users.

wilsonyqm replied:
make sense, removed

rmacklin commented on test/page_objects/images/signup_page.rb:
```diff
@@ -0,0 +1,23 @@
+module PageObjects
+  module Users
```
Similar to the above, I don't think we need this `Users` namespace.

wilsonyqm replied:
removed

rmacklin commented:
Just a few comments left. Good job!

wilsonyqm commented:
Addressed new comments, see https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36#discussion_r63226323

rmacklin commented on app/views/layouts/_navbar.html.erb:
```diff
@@ -1,13 +1,28 @@
+      <% else %>
+        <%= nav_item(text: 'Log In / Sign Up',
+                     path: login_path,
+                     is_active:  current_page?(login_path),
```
What if you are on the signup page?

wilsonyqm replied:
Right, is_active should be false in both of this cases.

rmacklin replied:
You mean `true`?

rmacklin commented on app/views/layouts/_navbar.html.erb:
```diff
@@ -1,13 +1,28 @@
+
+    <div class="pull-xs-right">
+      <% if @current_user.present? %>
+        <%= nav_item(text: 'Log Out',
```
`nav_item` generates `li` tags. These must be children of a `ul` (or `ol`) tag. Instead, they are children of

wilsonyqm replied:
Right, I find out where the problem is.

rmacklin replied:
You should just remove the `pull-xs-right` `div` and move this stuff into the above `ul`

rmacklin commented on app/views/layouts/_navbar.html.erb:
```diff
@@ -1,13 +1,28 @@
+        <%= nav_item(text: 'Log Out',
+                     path: logout_path,
+                     is_active:  current_page?(logout_path),
+                     class_list: ['pull-lg-left', 'fa', 'fa-sign-out'],
```
We've discussed this before, but font-awesome classes need to be on their own element, not on a container. The generated HTML should be:
```html
<li class="nav-item pull-xs-right">
  <a class="nav-link" data-method="get" href="/login"><span class="fa fa-sign-in"></span> Log In / Sign Up</a>
</li>
```
So if you would like to use a font-awesome icon correctly in the nav bar, you should make `nav_item` accept a block the same way that other helpers like `link_to` do (which you are using [here](https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/blob/ec636ff7ef1437566911aca700da076f056a9030/app/views/images/show.html.erb#L3-L9)). Alternatively, you can remove the icon.

rmacklin replied:
After this change, you should remove the `class_list` argument. Instead, you can have an `alignment` parameter which, if set to `:right` would add the `pull-xs-right` class to the `nav-item` element.

wilsonyqm replied:
Fixed.
```ruby
  def nav_item(text: nil, path:, is_active: true, class_list: {}, method: nil, alignment: 'left')
    classes = ['nav-item']
    if class_list.present?
      class_list.each { |klass| classes << klass }
    end
    classes << 'active' if is_active
    classes << "pull-xs-#{alignment}"
    content_tag :li, class: classes do
      if text.present?
        link_to text, path, class: 'nav-link', method: method
      else
        link_to path, class: 'nav-link', method: method do
          yield
        end
      end
    end
  end
```

rmacklin replied:
Looks pretty good, but I think you can remove the `class_list` argument since it was being used for font-awesome icon classes (incorrectly), and now we have the block handling for that. Additionally, I think you can also default `alignment` to `nil` and just write:
```ruby
classes << 'pull-xs-right' if alignment == :right
```
because `:right` is the only thing that should be passed.

rmacklin commented on app/helpers/application_helper.rb:
```diff
@@ -1,9 +1,12 @@
 module ApplicationHelper
-  def nav_item(text:, path:, is_active:)
+  def nav_item(text:, path:, is_active:, class_list: {}, method: :get)
```
As mentioned in https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36#discussion_r63248757, we should remove the `class_list` argument.

rmacklin replied:
Additionally, the way you've implemented this means that all normal links will have an unnecessary
`data-method="get"`
attribute. To avoid this, I think `method` should default to `nil` instead of `:get`.

wilsonyqm replied:
Fixed.

rmacklin commented on test/helpers/application_helper_test.rb:
```diff
@@ -3,13 +3,22 @@
   test 'nav_item' do
     expected_html = <<-HTML.strip
-<li class="nav-item active"><a class="nav-link" href="/images">Images</a></li>
+<li class=\"nav-item active\"><a class=\"nav-link\" data-method=\"get\" href=\"/images\">Images</a></li>
```
Why did you suddenly have to escape these quotes?

wilsonyqm replied:
I forgot to remove them, since I just copy paste from the test console.

rmacklin commented on db/migrate/20160426204118_create_users.rb:
```diff
@@ -1,4 +1,4 @@
-class CreateAndModifyUsers < ActiveRecord::Migration
+class CreateUsers < ActiveRecord::Migration
```
👍

rmacklin commented on test/mailers/image_mailer_test.rb:
```diff
@@ -80,7 +80,7 @@ class ImageMailerTest < ActionMailer::TestCase
 
-  test 'share image when logged in' do
+  test 'share image from a User' do
     user_logged_in = User.create!(
```
You should rename this variable too. Just `user` is fine.

wilsonyqm replied:
renamed

rmacklin commented:
I left some new comments, including https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36#discussion_r63249720 and https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36#discussion_r63246943 which appear "on an outdated diff" but are still applicable.

rmacklin commented on app/helpers/application_helper.rb:
```diff
@@ -1,9 +1,16 @@
 module ApplicationHelper
-  def nav_item(text:, path:, is_active:)
+  def nav_item(text: nil, path:, is_active: true, method: nil, alignment: 'left')
```
See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36#discussion_r63261726

wilsonyqm replied:
Modified

rmacklin commented on Gemfile:
```diff
@@ -62,8 +62,13 @@ gem 'acts-as-taggable-on', '~> 3.4'
 gem 'bootstrap', '>=4.0.0.alpha3'
 gem 'browserify-rails'
 gem 'simple_form'
+gem 'bcrypt', '3.1.7'
```
FYI, you could have uncommented line 26 instead

wilsonyqm replied:
changed

rmacklin commented on Gemfile:
```diff
@@ -62,8 +62,13 @@ gem 'acts-as-taggable-on', '~> 3.4'
+
+gem 'validates_email_format_of'
+
+gem 'mocha'
```
You should move this into the `test` group since it's only needed for tests.

rmacklin commented on app/views/layouts/_navbar.html.erb:
```diff
@@ -1,13 +1,29 @@
+        <% end %>
+      <% else %>
+        <%= nav_item(alignment: 'right',
+                     path: login_path) do %>
```
See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36#discussion_r63247848

rmacklin commented on app/views/layouts/_navbar.html.erb:
```diff
@@ -1,13 +1,29 @@
+
+    <div class="pull-xs-right">
+
+    </div>
```
See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36#discussion_r63259806

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -127,6 +127,26 @@ class ImagesControllerTest < ActionController::TestCase
+    password = 'passwd'
+    user = User.create!(email: username, password: password, name: 'name')
+    log_in(user)
+    something = ImageMailer.expects(:send_email).with(image1, anything, user).returns(mock(:deliver_now))
```
This assignment is unnecessary. Pronto should've complained about it.

wilsonyqm commented:
Address new comments. Thank you! :)

rmacklin commented on app/helpers/application_helper.rb:
```diff
@@ -1,9 +1,16 @@
+  def nav_item(text: nil, path:, is_active: true, method: nil, alignment: nil)
     classes = ['nav-item']
     classes << 'active' if is_active
+    alignment == :right ? classes << 'pull-xs-right' : classes << 'pull-xs-left'
```
You don't need to add the `pull-xs-left` class at all. By default, everything will be left-aligned. So line can just be:
`classes << 'pull-xs-right' if alignment == :right`

wilsonyqm replied:
Fixed.

rmacklin commented on app/views/layouts/_navbar.html.erb:
```diff
@@ -1,13 +1,25 @@
+        <% end %>
+      <% else %>
+        <%= nav_item(alignment: :right,
+                     path: login_path) do %>
```
How come you deleted the `is_active` attribute?

wilsonyqm replied:
Added it back.

rmacklin commented:
Almost done. Two comments left

wilsonyqm commented:
Fixed the bug mentioned above.

rmacklin commented:
Great! You can squash now.

wilsonyqm commented:
If it's good, can't wait to merge. 👍

rmacklin commented on test/page_objects/images/index_page.rb:
```diff
@@ -10,6 +10,16 @@ def view!
+      def log_out!
+        node.click_on('Log Out')
+        window.change_to(IndexPage)
+      end
```
As alluded to in https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/36#discussion_r62151922, do these methods belong on the `IndexPage`?

rmacklin commented on test/page_objects/document.rb:
```diff
@@ -3,5 +3,20 @@ class Document < AePageObjects::Document
+      window.change_to(Images::IndexPage)
+    end
+
+    def clear_tag_filter!
```
This method does not belong on `Document`. "Clearing the tag filter" is something that only applies to the `Images::IndexPage` (when it has been filtered by a tag)
